### PR TITLE
THRIFT-3765 fix memory leak in python compact protocol extension

### PR DESCRIPTION
**Pull request by cjmay**
This fixes a significant memory leak in the new Python extension module for the compact protocol.

### CODE REVIEW

**Comment by nsuke** [comment]
@cjmay thanks for catching this before the release.
[C extension + UTF8 fix](https://github.com/apache/thrift/pull/769/files#diff-cf2b2a6a8c345e5979b22a2ccc932abfR439) seems to be the true culprit here (so it's even much more worse in terms of the scope of this regression).

File: lib/py/src/ext/protocol.tcc [hunk 1]
```diff
@@ -431,6 +433,7 @@ bool ProtocolBase<Impl>::encodeValue(PyObject* value, TType type, PyObject* type
     }
```
Comment (nsuke) [inline lib/py/src/ext/protocol.tcc]: I think we need DECREF before the "return false" above too.
Feel free to use ScopedPyObject (or not).

Comment (cjmay) [inline lib/py/src/ext/protocol.tcc]: I thought so too, originally, but isn't value null there?

**Comment by cjmay** [comment]
Thank you!  We were excited to see this extension get written.

File: lib/py/src/ext/protocol.tcc (hunk 1 above)
Comment (nsuke) [inline lib/py/src/ext/protocol.tcc]: It shouldn't be...
The value is either null-checked or passed to `Py_INCREF` that does not accept null.
The intention here is to simply reject too long string values.
If you experienced null value here, we'll need to investigate on that matter too.

Comment (cjmay) [inline lib/py/src/ext/protocol.tcc]: Oh, I apologize, I was looking at the wrong "return false."

Comment (cjmay) [inline lib/py/src/ext/protocol.tcc]: Fixed, and thanks for the tip.

**Comment by nsuke** [comment]
Committed, thanks again for the fix.

**Pull request closed without merging**